File: 86-partition-list/partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode p = new ListNode(0);
        ListNode q = new ListNode(0);
        ListNode pp = p, qp = q;
        while(head != null){
            if (head.val < x){
                pp.next = head;
                pp = pp.next;
            }
            else if(head.val >= x){
                qp.next = head;
                qp = qp.next;
            }
            head = head.next;
        }
        pp.next = q.next;
        qp.next = null;
        return p.next;
    }
}

File: 86-partition-list/partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        d1 = ListNode(0)
        d2 = ListNode(0)
        p = d1
        q = d2
        while(head is not None):
            if head.val < x:
                p.next = head
                p = head
            elif head.val >= x:
                q.next = head
                q = head
            head = head.next
        q.next = None
        p.next = d2.next
        return d1.next